kbase-extension/static/kbase/js/common/jobManager.js

```diff
@@ -64,41 +121,157 @@ define(['common/html', 'common/jobs', 'common/ui', 'util/string'], (html, Jobs,
         }
 
         /**
-         * Remove an update handler
+         * Remove a handler
          *
+         * @param {string} event
          * @param {string} handlerName
          */
-        removeUpdateHandler(handlerName) {
-            const handlerFunction = this.handlers[handlerName];
-            delete this.handlers[handlerName];
-            return handlerFunction;
+        removeHandler(event, handlerName) {
+            if (this.handlers[event]) {
+                const handlerFunction = this.handlers[event][handlerName];
+                delete this.handlers[event][handlerName];
+                return handlerFunction;
+            }
         }
 
         /**
-         * Trigger the update handlers alphabetically by name.
+         * Trigger the ${event} handlers. Handlers are executed alphabetically by name.
          *
-         * @param {array} jobArray -- array of pertinent jobs
+         * By default, the handler functions receive the jobManager context as
+         * first argument and any additional arguments passed in by the caller
+         * (e.g. the message and job ID in the case of job listeners)
+         *
+         * @param {string} event
+         * @param  {...any} args
          */
-        runUpdateHandlers(jobArray) {
-            Object.keys(this.handlers)
+        runHandler(event, ...args) {
+            if (
+                !this._isValidEvent(event) ||
+                !this.handlers[event] ||
+                !Object.keys(this.handlers[event])
+            ) {
+                return;
+            }
+
+            Object.keys(this.handlers[event])
                 .sort()
                 .forEach((name) => {
                     try {
-                        this.handlers[name](this.model, jobArray);
+                        this.handlers[event][name](this, ...args);
                     } catch (err) {
                         console.warn(`Error executing handler ${name}:`, err);
                     }
                 });
         }
 
+        /* JOB LISTENERS */
+
+        /**
+         * Add a bus listener for ${event} messages
+         *
+         * @param {string} type - a valid message type to listen for (see validOutgoingMessageTypes)
+         * @param {array} jobIdList - array of job IDs to apply the listener to
+         * @param {object} handlerObject (optional) - object with key(s) handler name and value(s) function to execute on receiving a message
+         */
+        addListener(type, jobIdList, handlerObject) {
```

Comment:
Rather than having listeners in the various widgets, keep them in the jobManager and pass that around instead.